cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -81,91 +81,108 @@ def process_data(old_data, new_data)
-        changes.removed_stacks.each { |stack| puts pastel.red("- #{stack}") }
-        puts
+        changes.removed_stacks.each { |stack|
+          puts "  #{pastel.red('-')} #{stack}"
+          changes.removed_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.red('-')} #{s}"
+          end
+        }
       end
 
       unless changes.replaced_stacks.empty?
-        puts pastel.yellow("These stacks #{will} be replaced by other stacks:")
+        changes.replaced_stacks.each { |stack|
+          puts "  #{pastel.yellow('-/+')} #{stack}"
+          changes.upgraded_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.cyan('~')} #{s}"
+          end
+          changes.added_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.green('+')} #{s}"
+          end
+          changes.removed_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.red('-')} #{s}"
+          end
+        }
         changes.replaced_stacks.each do |installed_name, data|
           puts "- #{pastel.yellow(installed_name)} from #{pastel.cyan(data[:from])} to #{pastel.cyan(data[:to])}"
         end
         puts
       end
 
       unless changes.added_stacks.empty?
-        puts pastel.green("These new stack dependencies #{will} be installed:")
-        changes.added_stacks.each { |stack| puts pastel.green("- #{stack}") }
-        puts
+        changes.added_stacks.each { |stack|
+          puts "  #{pastel.green('+')} #{stack}"
+          changes.added_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.green('+')} #{s}"
+          end
+        }
       end
 
       unless changes.upgraded_stacks.empty?
-        puts pastel.cyan("These stacks #{will} be upgraded#{' and deployed' if deploy?}:")
-        changes.upgraded_stacks.each { |stack| puts pastel.cyan("- #{stack}") }
-        puts
+        changes.upgraded_stacks.each { |stack|
+          puts "  #{pastel.cyan('~')} #{stack}"
+          changes.upgraded_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.cyan('~')} #{s}"
+          end
+          changes.added_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.green('+')} #{s}"
+          end
+          changes.removed_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.red('-')} #{s}"
+          end
+        }
       end
-
       puts
     end
 
     # requires heavier confirmation when something very dangerous is going to happen
     def get_confirmation(changes)
       unless force?
-        unless changes.removed_services.empty? && changes.removed_stacks.empty? && changes.replaced_stacks.empty?
+        unless changes.safe?
           puts "#{pastel.red('Warning:')} This can not be undone, data will be lost."
           confirm
```

Comment:
(these could be `confirm("This can't be undone, data will be lost.") unless safe?`)